cpp/src/arrow/compute/kernels/vector_swizzle_test.cc

```diff
@@ -0,0 +1,1065 @@
+  DCHECK(if_true_result_datum.is_array());
+  auto if_true_result_array = if_true_result_datum.make_array();
+
+  // 7. Evaluate "false" branch on the "false" rows.
+  ARROW_ASSIGN_OR_RAISE(auto bound_if_false, if_false.Bind(*schema));
+  ARROW_ASSIGN_OR_RAISE(auto if_false_result_datum,
+                        ExecuteScalarExpression(bound_if_false, if_false_input_batch));
+  DCHECK(if_false_result_datum.is_array());
+  auto if_false_result_array = if_false_result_datum.make_array();
+
+  // 8. Combine the "true"/"false" results/selection vectors into chunked arrays.
+  auto result_ca = std::make_shared<ChunkedArray>(
+      ArrayVector{if_true_result_array, if_false_result_array});
+  auto sel_ca =
+      std::make_shared<ChunkedArray>(ArrayVector{sel_if_true_array, sel_if_false_array});
+
+  // 9. Finally, scatter the "true"/"false" results to their original positions in the
+  // input (according to the selection vectors). Note we didn't handle the rows with nulls
+  // in the mask, because Scatter will fill nulls for these rows and this is equal to the
+  // null handling policy of if-else, which is pretty nice.
+  return Scatter(/*values=*/result_ca, /*indices=*/sel_ca,
+                 /*max_index=*/input.length - 1);
+}
+
+void DoTestIfElse(const Expression& cond, const Expression& if_true,
+                  const Expression& if_false, const std::shared_ptr<Schema>& schema,
+                  const ExecBatch& input) {
+  ASSERT_OK_AND_ASSIGN(Datum result_by_expr,
+                       ExecuteIfElseByExpr(cond, if_true, if_false, schema, input));
+  ASSERT_TRUE(result_by_expr.is_array());
+  ASSERT_OK_AND_ASSIGN(Datum result_by_scatter,
+                       ExecuteIfElseByScatter(cond, if_true, if_false, schema, input));
+  // Scatter will output chunked array because we input values and indices as chunked
+  // arrays consisting of each branches. We don't care the shape of the output when
+  // comparing the results - only contents, so we concatenate the chunked array.
+  ASSERT_TRUE(result_by_scatter.is_chunked_array());
+  ASSERT_OK_AND_ASSIGN(auto result_by_scatter_concat,
+                       Concatenate(result_by_scatter.chunked_array()->chunks()));
+
+  AssertDatumsEqual(result_by_expr, result_by_scatter_concat);
+}
+
+void DoTestIfElse(const Expression& cond, const Expression& if_true,
+                  const Expression& if_false, const std::shared_ptr<Schema>& schema,
+                  const ExecBatch& input, const std::shared_ptr<Array>& expected) {
+  ASSERT_OK_AND_ASSIGN(Datum result,
+                       ExecuteIfElseByScatter(cond, if_true, if_false, schema, input));
+  ASSERT_TRUE(result.is_chunked_array());
+  ASSERT_OK_AND_ASSIGN(auto result_concat, Concatenate(result.chunked_array()->chunks()));
+
+  AssertDatumsEqual(expected, result_concat);
+}
+
+}  // namespace
+
+TEST(Scatter, IfElse) {
+  {
+    ARROW_SCOPED_TRACE("if (b != 0) then a / b else b");
```

Comment:
Are these tests really useful to validate the behavior of the scatter function?

Reply:
This is more of a show-case of how scatter is utilized to implement "if-else" special form.

I would say this isn't a typical UT that is specialized for a particular aspect of the function itself. But it does make sure the function is designed and implemented as what it is expected to work with another upcoming feature (special form), and, of course, exercise the function from an end-to-end real world perspective.

Once the special form was implemented, we can retire this test case. What do you think?

Reply:
Ok, then:
1) can we make the test minimal?
2) add a comment explaining why the tests, and perhaps a TODO to remove them later?

Reply:
OK. Since I have my special form implementation almost done, I'm pretty sure the scatter function serves it well. So I decide to remove this case.